hw/ip/otbn/dv/rig/rig/gens/bad_load_store.py

```diff
@@ -0,0 +1,284 @@
+
+        # LW has three operands: grd, offset, grs
+        if not (len(self.lw.operands) == 3 and
+                isinstance(self.lw.operands[0].op_type, RegOperandType) and
+                self.lw.operands[0].op_type.reg_type == 'gpr' and
+                self.lw.operands[0].op_type.is_dest() and
+                isinstance(self.lw.operands[1].op_type, ImmOperandType) and
+                isinstance(self.lw.operands[2].op_type, RegOperandType) and
+                self.lw.operands[2].op_type.reg_type == 'gpr' and
+                not self.lw.operands[2].op_type.is_dest()):
+            raise RuntimeError('LW instruction from instructions file is '
+                               'not the shape expected by the BadLoadStore'
+                               'generator.')
+
+        grd_op_type = self.lw.operands[0].op_type
+        imm_op_type = self.lw.operands[1].op_type
+
+        assert imm_op_type.shift == 0
+
+        # SW Instruction Checks
+        # SW has three operands: grs1, offset, grs2
+        if not (len(self.sw.operands) == 3 and
+                isinstance(self.sw.operands[0].op_type, RegOperandType) and
+                self.sw.operands[0].op_type.reg_type == 'gpr' and
+                not self.sw.operands[0].op_type.is_dest() and
+                isinstance(self.sw.operands[1].op_type, ImmOperandType) and
+                isinstance(self.sw.operands[2].op_type, RegOperandType) and
+                self.sw.operands[2].op_type.reg_type == 'gpr' and
+                not self.sw.operands[2].op_type.is_dest()):
+            raise RuntimeError('SW instruction from instructions file is '
+                               'not the shape expected by the BadLoadStore'
+                               'generator.')
+
+        # Determine the offset range for XW
+        offset_rng = imm_op_type.get_op_val_range(model.pc)
+        assert offset_rng is not None
+
+        # Max/Min Offsets for XW (-2048, 2047)
+        min_offset, max_offset = offset_rng
+
+        # Get known registers
+        known_regs = model.regs_with_known_vals('gpr')
+
+        # Pick a random register among known registers
+        idx, value = random.choice(known_regs)
+
+        # If it is close enough, pick top of the memory for "barely" out of bound
+        # coverage point for XW instructions
+        if (model.dmem_size + min_offset < value and  # value > 4096 + (-2048)
+           value <= model.dmem_size - min_offset):  # value <= 4096 - (-2048)
+            imm_val = model.dmem_size - value
+        else:
+            imm_val = random.randrange(min_offset, max_offset)
+
+        if value >> 31:
+            addr = imm_val + (value - (1 << 32))
+        else:
+            addr = imm_val + value
+
+        while not (addr < 0 or model.dmem_size <= addr):
+            imm_val = random.randrange(min_offset, max_offset)
+            if value >> 31:
+                addr = imm_val + (value - (1 << 32))
+            else:
+                addr = imm_val + value
```

Comment:
I'm really sorry, but I still don't understand what this logic is supposed to be doing. Could you rewrite it to look "more like a constraint solver"? I'm thinking of something like:
- Pick a base register
- Check whether it's possible to hit the first address after the top of memory and, if so, whether we want to
- Pick an address to hit from the range of representable addresses
- Compute the required value of `imm_val` accordingly.

I really think that this would be much easier to understand and (I think) it would also be easier to control how often the "edge" case happens too.

Similarly for `_fill_bn_xid` below.

Reply:
Thanks for the review and suggestions @rswarbrick. I tried to make it more clear, I am sorry about the low quality code.